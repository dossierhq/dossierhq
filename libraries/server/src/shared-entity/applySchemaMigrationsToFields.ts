import {
  isComponentItemField,
  isRichTextComponentNode,
  ok,
  transformEntityFields,
  type Component,
  type ErrorType,
  type Result,
  type SchemaMigrationAction,
  type SchemaWithMigrations,
} from '@dossierhq/core';
import type { DatabaseEntityFieldsPayload } from '@dossierhq/database-adapter';
import { assertExhaustive } from '../utils/AssertUtils.js';

export function applySchemaMigrationsToFields(
  schema: SchemaWithMigrations,
  targetEntityType: string,
  entityFields: DatabaseEntityFieldsPayload,
): Result<Record<string, unknown>, typeof ErrorType.BadRequest | typeof ErrorType.Generic> {
  const actions = schema.collectMigrationActionsSinceVersion(entityFields.schemaVersion);

  if (actions.length === 0) {
    return ok(entityFields.fields);
  }

  const entityTypeActions: Exclude<SchemaMigrationAction, { componentType: string }>[] = [];
  const componentTypeActions: Exclude<SchemaMigrationAction, { entityType: string }>[] = [];
  for (const action of actions) {
    if ('entityType' in action) {
      entityTypeActions.push(action);
    } else {
      componentTypeActions.push(action);
    }
  }

  const startingEntityType = getStartingEntityType(targetEntityType, entityTypeActions);

  const migratedFieldValues = migrateEntityFields(
    startingEntityType,
    entityFields.fields,
    entityTypeActions,
  );

  const transformResult: Result<
    Record<string, unknown>,
    typeof ErrorType.BadRequest | typeof ErrorType.Generic
  > = transformEntityFields(
    schema,
    [],
    { info: { type: targetEntityType }, fields: migratedFieldValues },
    {
      transformField: (_schema, _path, _fieldSpec, value) => {
        return ok(value);
      },
      transformFieldItem: (_schema, _path, fieldSpec, value) => {
        if (isComponentItemField(fieldSpec, value) && value) {
          return ok(migrateComponent(value, componentTypeActions));
        }
        return ok(value);
      },
      transformRichTextNode: (_schema, _path, _fieldSpec, node) => {
        if (isRichTextComponentNode(node)) {
          const component = migrateComponent(node.data, componentTypeActions);
          if (!component) return ok(null);
          return ok({ ...node, data: component });
        }
        return ok(node);
      },
    },
  );

  if (transformResult.isError()) return transformResult;
  return ok(transformResult.value);
}

function getStartingEntityType(
  targetEntityType: string,
  entityTypeActions: Exclude<SchemaMigrationAction, { componentType: string }>[],
) {
  let entityType = targetEntityType;
  for (let i = entityTypeActions.length - 1; i >= 0; i--) {
    const action = entityTypeActions[i];
    if (action.action === 'renameType' && action.newName === entityType) {
      entityType = action.entityType;
    }
  }
  return entityType;
}

function migrateEntityFields(
  startingEntityType: string,
  originalFields: Record<string, unknown>,
  entityTypeActions: Exclude<SchemaMigrationAction, { componentType: string }>[],
) {
  let changed = false;
  let entityType = startingEntityType;
  const migratedFields = { ...originalFields };
  for (const actionSpec of entityTypeActions) {
    const { action } = actionSpec;
    switch (action) {
      case 'deleteField':
        if (actionSpec.entityType === entityType) {
          delete migratedFields[actionSpec.field];
          changed = true;
        }
        break;
      case 'renameField':
        if (actionSpec.entityType === entityType) {
          if (actionSpec.field in migratedFields) {
            migratedFields[actionSpec.newName] = migratedFields[actionSpec.field];
            delete migratedFields[actionSpec.field];
            changed = true;
          }
        }
        break;
      case 'renameType':
        if (actionSpec.entityType === entityType) {
          entityType = actionSpec.newName;
        }
        break;
    }
  }
  return changed ? migratedFields : originalFields;
}

function migrateComponent(
  originalComponent: Component | null,
  componentTypeActions: Exclude<SchemaMigrationAction, { entityType: string }>[],
): Component | null {
  if (!originalComponent) return null;

  const component = { ...originalComponent };

  for (const actionSpec of componentTypeActions) {
    const { action } = actionSpec;
    switch (action) {
      case 'deleteField':
        if (actionSpec.componentType === component.type) {
          delete component[actionSpec.field];
        }
        break;
      case 'renameField':
        if (actionSpec.componentType === component.type) {
          if (actionSpec.field in component) {
            component[actionSpec.newName] = component[actionSpec.field];
            delete component[actionSpec.field];
          }
        }
        break;
      case 'deleteType':
        if (actionSpec.componentType === component.type) {
          return null;
        }
        break;
      case 'renameType':
        if (actionSpec.componentType === component.type) {
          component.type = actionSpec.newName;
        }
        break;
      default:
        assertExhaustive(action);
    }
  }
  return component;
}
