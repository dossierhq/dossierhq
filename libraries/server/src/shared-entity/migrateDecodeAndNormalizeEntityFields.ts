import {
  isComponentItemField,
  isRichTextComponentNode,
  notOk,
  ok,
  transformEntityFields,
  type Entity,
  type EntityTypeSpecification,
  type SchemaWithMigrations,
  type ContentTransformer,
  type ContentValuePath,
  type ErrorType,
  type PublishedEntityTypeSpecification,
  type PublishedSchema,
  type Result,
} from '@dossierhq/core';
import type { DatabaseEntityFieldsPayload } from '@dossierhq/database-adapter';
import { applySchemaMigrationsToFields } from './applySchemaMigrationsToFields.js';
import { legacyDecodeEntityFields } from './legacyDecodeEntityFields.js';

export const ENCODE_VERSION_LEGACY = 0;
export const ENCODE_VERSION_AS_IS = 1; // version 0.4.2+: Used to encoding all content, old content is still encoded with version 0

export function migrateDecodeAndNormalizeAdminEntityFields(
  schema: SchemaWithMigrations,
  entitySpec: EntityTypeSpecification,
  path: ContentValuePath,
  entityFields: DatabaseEntityFieldsPayload,
): Result<Entity['fields'], typeof ErrorType.BadRequest | typeof ErrorType.Generic> {
  return migrateDecodeAndNormalizeEntityFields(schema, schema, entitySpec, path, entityFields);
}

export function migrateDecodeAndNormalizePublishedEntityFields(
  schema: SchemaWithMigrations,
  entitySpec: PublishedEntityTypeSpecification,
  path: ContentValuePath,
  entityFields: DatabaseEntityFieldsPayload,
): Result<Entity['fields'], typeof ErrorType.BadRequest | typeof ErrorType.Generic> {
  return migrateDecodeAndNormalizeEntityFields(
    schema,
    schema.toPublishedSchema(),
    entitySpec,
    path,
    entityFields,
  );
}

function migrateDecodeAndNormalizeEntityFields<
  TSchema extends SchemaWithMigrations | PublishedSchema,
>(
  schema: SchemaWithMigrations,
  decodeSchema: TSchema,
  entitySpec: TSchema['spec']['entityTypes'][number],
  path: ContentValuePath,
  entityFields: DatabaseEntityFieldsPayload,
): Result<Entity['fields'], typeof ErrorType.BadRequest | typeof ErrorType.Generic> {
  const migratedFieldValuesResult = applySchemaMigrationsToFields(
    schema,
    entitySpec.name,
    entityFields,
  );
  if (migratedFieldValuesResult.isError()) return migratedFieldValuesResult;
  const migratedFieldValues = migratedFieldValuesResult.value;

  switch (entityFields.encodeVersion) {
    case ENCODE_VERSION_LEGACY:
      return ok(legacyDecodeEntityFields(decodeSchema, entitySpec, migratedFieldValues));
    case ENCODE_VERSION_AS_IS: {
      const normalizeResult = transformEntityFields(
        decodeSchema,
        path,
        { info: { type: entitySpec.name }, fields: migratedFieldValues },
        DECODE_TRANSFORMER,
      );
      if (normalizeResult.isError()) return normalizeResult;
      return ok(normalizeResult.value);
    }
    default:
      return notOk.Generic(`Unknown encode version ${entityFields.encodeVersion}`);
  }
}

const DECODE_TRANSFORMER: ContentTransformer<
  SchemaWithMigrations | PublishedSchema,
  typeof ErrorType.Generic
> = {
  transformField(_schema, _path, _fieldSpec, value: unknown) {
    return ok(value);
  },
  transformFieldItem(schema, _path, fieldSpec, value: unknown) {
    if (isComponentItemField(fieldSpec, value)) {
      const componentType = value?.type;
      if (componentType && !schema.getComponentTypeSpecification(componentType)) {
        // Could be that the component type was deleted or made adminOnly (when decoding published entities)
        return ok(null);
      }
    }
    return ok(value);
  },
  transformRichTextNode(schema, _path, _fieldSpec, node) {
    if (isRichTextComponentNode(node)) {
      const componentType = node.data.type;
      if (componentType && !schema.getComponentTypeSpecification(componentType)) {
        // Could be that the component type was deleted or made adminOnly (when decoding published entities)
        return ok(null);
      }
    }
    return ok(node);
  },
};
