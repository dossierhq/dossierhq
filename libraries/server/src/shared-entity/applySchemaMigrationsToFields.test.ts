import {
  AdminSchemaWithMigrations,
  FieldType,
  createRichTextHeadingNode,
  createRichText,
  createRichTextTextNode,
  createRichTextValueItemNode,
  type RichText,
  type RichTextValueItemNode,
} from '@dossierhq/core';
import { describe, expect, test } from 'vitest';
import { applySchemaMigrationsToFields } from './applySchemaMigrationsToFields.js';
import { ENCODE_VERSION_AS_IS } from './migrateDecodeAndNormalizeEntityFields.js';

const ADMIN_SCHEMA_BASE = AdminSchemaWithMigrations.createAndValidate({
  entityTypes: [
    {
      name: 'Entity',
      fields: [
        { name: 'richText', type: FieldType.RichText },
        { name: 'valueItem', type: FieldType.ValueItem },
        { name: 'valueItemList', type: FieldType.ValueItem, list: true },
      ],
    },
  ],
  valueTypes: [
    {
      name: 'ValueItem',
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'child', type: FieldType.ValueItem },
      ],
    },
  ],
}).valueOrThrow();

describe('applySchemaMigrationsToFields renameField', () => {
  test('nested value item', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [
            {
              action: 'renameField',
              valueType: 'ValueItem',
              field: 'string',
              newName: 'string2',
            },
            {
              action: 'renameField',
              valueType: 'ValueItem',
              field: 'child',
              newName: 'child2',
            },
          ],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItem: {
          type: 'ValueItem',
          string: '1',
          child: {
            type: 'ValueItem',
            string: '1.1',
            child: { type: 'ValueItem', string: '1.1.1', child: null },
          },
        },
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchInlineSnapshot(`
      {
        "richText": null,
        "valueItem": {
          "child2": {
            "child2": {
              "child2": null,
              "string2": "1.1.1",
              "type": "ValueItem",
            },
            "string2": "1.1",
            "type": "ValueItem",
          },
          "string2": "1",
          "type": "ValueItem",
        },
        "valueItemList": null,
      }
    `);
  });

  test('value item in rich text', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [
            {
              action: 'renameField',
              valueType: 'ValueItem',
              field: 'string',
              newName: 'string2',
            },
            {
              action: 'renameField',
              valueType: 'ValueItem',
              field: 'child',
              newName: 'child2',
            },
          ],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        richText: createRichText([
          createRichTextHeadingNode('h1', [createRichTextTextNode('Heading 1')]),
          createRichTextValueItemNode({
            type: 'ValueItem',
            string: '1',
            child: { type: 'ValueItem', string: '1.1', child: null },
          }),
        ]),
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
    expect(((fieldValues.richText as RichText).root.children[1] as RichTextValueItemNode).data)
      .toMatchInlineSnapshot(`
      {
        "child2": {
          "child2": null,
          "string2": "1.1",
          "type": "ValueItem",
        },
        "string2": "1",
        "type": "ValueItem",
      }
    `);
  });
});

describe('applySchemaMigrationsToFields deleteField', () => {
  test('nested value item', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [{ action: 'deleteField', valueType: 'ValueItem', field: 'string' }],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        richText: createRichText([
          createRichTextHeadingNode('h1', [createRichTextTextNode('Heading 1')]),
          createRichTextValueItemNode({
            type: 'ValueItem',
            string: '1',
            child: { type: 'ValueItem', string: '1.1', child: null },
          }),
        ]),
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
    expect(((fieldValues.richText as RichText).root.children[1] as RichTextValueItemNode).data)
      .toMatchInlineSnapshot(`
        {
          "child": {
            "child": null,
            "type": "ValueItem",
          },
          "type": "ValueItem",
        }
      `);
  });

  test('value item in rich text', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [{ action: 'deleteField', valueType: 'ValueItem', field: 'string' }],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItem: {
          type: 'ValueItem',
          string: '1',
          child: {
            type: 'ValueItem',
            string: '1.1',
            child: { type: 'ValueItem', string: '1.1.1', child: null },
          },
        },
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchInlineSnapshot(`
      {
        "richText": null,
        "valueItem": {
          "child": {
            "child": {
              "child": null,
              "type": "ValueItem",
            },
            "type": "ValueItem",
          },
          "type": "ValueItem",
        },
        "valueItemList": null,
      }
    `);
  });
});

describe('applySchemaMigrationsToFields renameType', () => {
  test('nested value item', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [{ action: 'renameType', valueType: 'ValueItem', newName: 'ValueItem2' }],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItem: {
          type: 'ValueItem',
          string: '1',
          child: {
            type: 'ValueItem',
            string: '1.1',
            child: { type: 'ValueItem', string: '1.1.1', child: null },
          },
        },
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchInlineSnapshot(`
      {
        "richText": null,
        "valueItem": {
          "child": {
            "child": {
              "child": null,
              "string": "1.1.1",
              "type": "ValueItem2",
            },
            "string": "1.1",
            "type": "ValueItem2",
          },
          "string": "1",
          "type": "ValueItem2",
        },
        "valueItemList": null,
      }
    `);
  });

  test('value item list', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [{ action: 'renameType', valueType: 'ValueItem', newName: 'ValueItem2' }],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItemList: [{ type: 'ValueItem', string: '1', child: null }],
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchInlineSnapshot(`
      {
        "richText": null,
        "valueItem": null,
        "valueItemList": [
          {
            "child": null,
            "string": "1",
            "type": "ValueItem2",
          },
        ],
      }
    `);
  });

  test('value item in rich text', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [{ action: 'renameType', valueType: 'ValueItem', newName: 'ValueItem2' }],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        richText: createRichText([
          createRichTextHeadingNode('h1', [createRichTextTextNode('Heading 1')]),
          createRichTextValueItemNode({
            type: 'ValueItem',
            string: '1',
            child: { type: 'ValueItem', string: '1.1', child: null },
          }),
        ]),
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
    expect(((fieldValues.richText as RichText).root.children[1] as RichTextValueItemNode).data)
      .toMatchInlineSnapshot(`
      {
        "child": {
          "child": null,
          "string": "1.1",
          "type": "ValueItem2",
        },
        "string": "1",
        "type": "ValueItem2",
      }
    `);
  });
});

describe('applySchemaMigrationsToFields deleteType', () => {
  test('nested value item', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [{ version: 2, actions: [{ action: 'deleteType', valueType: 'ValueItem' }] }],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItem: {
          type: 'ValueItem',
          string: '1',
          child: {
            type: 'ValueItem',
            string: '1.1',
            child: { type: 'ValueItem', string: '1.1.1', child: null },
          },
        },
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
  });

  test('value item list', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [{ version: 2, actions: [{ action: 'deleteType', valueType: 'ValueItem' }] }],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItemList: [{ type: 'ValueItem', string: '1', child: null }],
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
  });

  test('value item in rich text', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [{ version: 2, actions: [{ action: 'deleteType', valueType: 'ValueItem' }] }],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        richText: createRichText([
          createRichTextHeadingNode('h1', [createRichTextTextNode('Heading 1')]),
          createRichTextValueItemNode({
            type: 'ValueItem',
            string: '1',
            child: { type: 'ValueItem', string: '1.1', child: null },
          }),
        ]),
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchSnapshot();
  });
});

describe('applySchemaMigrationsToFields combos', () => {
  test('rename type and field', () => {
    const adminSchema = ADMIN_SCHEMA_BASE.updateAndValidate({
      migrations: [
        {
          version: 2,
          actions: [
            { action: 'renameType', valueType: 'ValueItem', newName: 'ValueItem2' },
            { action: 'renameField', valueType: 'ValueItem2', field: 'string', newName: 'string2' },
          ],
        },
      ],
    }).valueOrThrow();

    const fieldValues = applySchemaMigrationsToFields(adminSchema, 'Entity', {
      schemaVersion: 1,
      encodeVersion: ENCODE_VERSION_AS_IS,
      fields: {
        valueItem: {
          type: 'ValueItem',
          string: '1',
          child: {
            type: 'ValueItem',
            string: '1.1',
            child: { type: 'ValueItem', string: '1.1.1', child: null },
          },
        },
      },
    }).valueOrThrow();
    expect(fieldValues).toMatchInlineSnapshot(`
      {
        "richText": null,
        "valueItem": {
          "child": {
            "child": {
              "child": null,
              "string2": "1.1.1",
              "type": "ValueItem2",
            },
            "string2": "1.1",
            "type": "ValueItem2",
          },
          "string2": "1",
          "type": "ValueItem2",
        },
        "valueItemList": null,
      }
    `);
  });
});
