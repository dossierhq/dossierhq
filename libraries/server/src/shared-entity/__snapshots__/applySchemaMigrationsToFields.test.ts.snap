// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applySchemaMigrationsToFields deleteField > nested value item 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child": {
              "child": null,
              "type": "ValueItem",
            },
            "type": "ValueItem",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`applySchemaMigrationsToFields deleteType > nested value item 1`] = `
{
  "richText": null,
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`applySchemaMigrationsToFields deleteType > value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`applySchemaMigrationsToFields deleteType > value item list 1`] = `
{
  "richText": null,
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`applySchemaMigrationsToFields renameField > value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child2": {
              "child2": null,
              "string2": "1.1",
              "type": "ValueItem",
            },
            "string2": "1",
            "type": "ValueItem",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`applySchemaMigrationsToFields renameType > value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child": {
              "child": null,
              "string": "1.1",
              "type": "ValueItem2",
            },
            "string": "1",
            "type": "ValueItem2",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
  "valueItem": null,
  "valueItemList": null,
}
`;
